### ISSUE TITLE
ofsoftswitch13 tcp bottleneck?

### ISSUE THREAD

**Issue by jm-boley** [CPqD/ofsoftswitch13#191]
Before I start, just let me say that I'm VERY new to networking and SDN in general, so please excuse any questions that may seem ignorant or uninformed. I'm just a student, so if I'm missing anything obvious don't hesitate to point it out, I probably don't know.

We're working with ofsoftswitch running on Mininet with the floodlight controller, deployed on a 72-core RHEL7 server box (no VM). At first blush everything appears to work as it should (at least, no connectivity issues), but when we test TCP transfers with iperf we're seeing a bottleneck somewhere that consistently limits throughput to ~500K - 1Mbit/s range on a 1Gbit simulated link. We've tried tuning the host's TCP settings (increasing TCP window size, etc) and forcing iperf to create larger buffers, without much impact on the iperf data. It seems safe to say that our throughput problems aren't due to buffer overruns on the simulated hosts.

Unless I'm mistaken, a large number of packet drops will cause TCP to back off on how much bandwidth it tries to use, and we're seeing a fairly significant number of retries in the iperf data and Cwnd sizes are consistently around 5K. What's odd is that UDP throughput is very close to 100% link utilization, with no appreciable packet loss. Also, we're seeing near 100% utilization of a core running the ofdatapath process during UDP trials, but the highest we've seen with TCP is 5%.

I'm tempted to think that ofsoftswitch13 doesn't play well with Mininet, or vice versa. If anyone has had any comparable experiences and figured out a solution I'd love to hear about it! Any input at all would be appreciated.

Thanks!

**Comment by jm-boley** [comment]
Just a followup, in case anyone else encounters this problem.

After the original post I started noticing a lot of error messages in /var/log/messages (/var/log/syslog on debian-based distros) of the following forms:

```
ofdatapath: 00027|netdev|WARN|Dropped <x> messages in last <y> seconds due to excessive rate
ofdatapath: 00028|netdev|WARN|error sending Ethernet packet on <interface>: Message too long
```

The only way to fix this problem is to turn off segment offloading on all affected interfaces. In a Mininet environment, that means using ethtool on every virtual node, including the hosts. From the Mininet prompt:

```
<node> ethtool --offload <interface> gso off tso off gro off
```

To all appearances CPqD softswitch, either on its own or in concert with Mininet, can't handle modern linux kernel packet processing features like GSO and GRO. I suspect the problem lies with CPqD softswitch in light of the fact that ovs (on the same box, running in the same environment) has no comparable problems to speak of, but it's possible that this switch simply doesn't play well with Mininet.

**Comment by ederlf** [comment]
Please, can you tell me which is the last commit of the ofsoftswitch you're
using?

**Comment by jm-boley** [comment]
I'm using the latest commit (8d3df820f).

A quick update, I've dug a bit more and played around with the settings, and it looks like TSO is the major culprit.

**Comment by ederlf** [comment]
Yes, it is a known problem.(yes, it should be written somewhere, but long time without checking issues here) Although I fixed the crashing problem caused previously, TSO still degrades performance.

**Comment by Anees-alnajjar** [comment]
Hi 
I am using Ofdatapath 1.3 and Ryu controller.
When I send small TCP requests to the server, it successes and there is problem with CPqD switch. However, If I increase the number of TCP requests, the switch crashes and I have to restart my topology again. And, when I check the /var/log/syslog file I found the next error:
"ofdatapath: 00008|netdev|WARN|error sending Ethernet packet on veth0: Message too long"
I tried "ethtool --offload <interface> gso off tso off gro off" to all the switch ports, including veth0. However, I have got the same problem.

What should I do?
Anees
